src/store/mod.rs

```diff
@@ -0,0 +1,190 @@
+// This file is part of Astarte.
+//
+// Copyright 2021 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Provides functionality for instantiating an Astarte sqlite database.
+
+use std::{error::Error as StdError, fmt::Debug};
+
+use async_trait::async_trait;
+
+pub use self::sqlite::SqliteStore;
+use crate::types::AstarteType;
+
+pub mod error;
+pub mod memory;
+pub mod sqlite;
+pub mod wrapper;
+
+/// Trait providing compatibility with Astarte devices to databases.
+///
+/// Any database implementing this trait can be used as permanent storage for the properties
+/// of an Astarte device.
+///
+/// This SDK provides an implementation of a sqlite database for which this trait has already
+/// been implemented, see [`crate::store::sqlite::SqliteStore`].
+#[async_trait]
+pub trait PropertyStore: Debug + Send + Sync + 'static
+where
+    // NOTE: the 'static bound is required for the MSRV, in other version the error is not present
+    Self::Err: StdError,
+{
+    type Err;
+
+    /// Stores a property within the database.
+    async fn store_prop(
+        &self,
+        interface: &str,
+        path: &str,
+        value: &AstarteType,
+        interface_major: i32,
+    ) -> Result<(), Self::Err>;
+    /// Load a property from the database.
```

Comment:
I would maybe specify in this docs that it might also delete the property from the database if a version mismatch is detected.

Reply:
That is implementation specific, I will add a SHOULD